README.md
# WireSep

## Privilege separated implementation of WireGuard for OpenBSD

Features:
* Privilege separation, long-term secrets in a separate process
* Short-term secrets separated per tunnel interface
* Fork/re-exec for a different memory layout between different networks
* Tight [pledge(2)] and no file system access after startup
* Easy to read and write config file format for humans, easy to parse for
  machines (small attack surface)
* Uses connected sockets for performance and reliability

Status: **beta**

## Requirements

* OpenBSD 6.4 or higher

## Install

```sh
$ doas pkg_add wiresep
```

Generate a new private key for the `tun0` interface.
```sh
$ doas wiresep-keygen tun0
```

Then create a configuration and store it in */etc/wiresep/wiresep.conf*. A
simple example looks like the following:

```
# This is an example of a server listening on the public ip 198.51.100.7 port
# 1022. It uses the tun0 device with the internal ip addresses 2001:db8::7
# and 172.16.0.1 and allows communication with the peer Jane and Joe. Jane is
# allowed to use any source ip, while Joe may only use 2001:db8::4 or
# 172.16.0.11/30 as the source ip of his packets. The private key for the tun0
# interface can be generated with `wiresep-keygen tun0`.

interface tun0 {
	ifaddr 2001:db8::7/126
	ifaddr 172.16.0.1/24

	listen 198.51.100.7:1022

	peer jane {
		pubkey BhyBpDfD7joIPPpjBW/g/Wdhiu3iVOzQhKodbsLqJ3A=
		allowedips *
	}

	peer joe {
		pubkey AhyBpDfD7joIPPpjBW/g/Wdhiu3iVOzQhKodbsLqJ3A=
		allowedips 2001:db8::4
		allowedips 172.16.0.11/30
	}
}
```
See [wiresep.conf(5)] for a complete description of the configuration file.

Once everyting is set, run [wiresep(8)]:

```sh
$ doas wiresep
```

## Documentation

Refer to the manuals for documentation and a configuration example:
* [wiresep-keygen(1)]
* [wiresep.conf(5)]
* [wiresep(8)]

The design documents can be found in the [doc](doc/) directory.

## Todo

* WireGuard Cookie support

## Known issues

* Interface aliases are not supported for outbound connections. In order to
  properly support local address selection without having to rely on an
  unpledged superuser process a new system call that would combine bind(2) and
  connect(2) would be required.

## License

ISC

Copyright (c) 2018, 2019, 2020 Tim Kuijsten

Permission to use, copy, modify, and distribute this software for any purpose
with or without fee is hereby granted, provided that the above copyright notice
and this permission notice appear in all copies.

THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY AND
FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM LOSS
OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER
TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR PERFORMANCE OF
THIS SOFTWARE.

---

[pledge(2)]: http://man.openbsd.org/pledge
[wiresep-keygen(1)]: https://netsend.nl/wiresep/wiresep-keygen.1.html
[wiresep.conf(5)]: https://netsend.nl/wiresep/wiresep.conf.5.html
[wiresep(8)]: https://netsend.nl/wiresep/wiresep.8.html


ChangeLog
2020-04-07, Version 0.11.1

* move away from GitHub to our own controlled domain, the new website can be
  found at https://netsend.nl/wiresep and the repository can be cloned at
  https://netsend.nl/wiresep.git
* support stable releases and distribute them through
  https://netsend.nl/wiresep/archive
* minor nitpicks


2020-04-07, Version 0.11.0

* let wiresep-keygen(1) securely store keys in a file
* change default file extension of private keys from .key to .privkey
* fix initiator role on big-endian systems
* fix a crash because of a memory misalignment on sparc64
* fix premature exit with multiple configured interfaces
* various improvements to the documentation
* improve log messages seen with -v and -vv
* lower tunnel mtu from 1420 to 1408 to work better with DSL


2020-03-25, Version 0.10.1

* upstream OpenBSD package patches that fix some build issues on base-gcc
  (32-bit) archs, patches thanks to Theo Buehler


2020-03-25, Version 0.10.0

* major effort to clarify all log messages
* fix all warnings and false positives reported by Clang Static Analyzer
* fix some warnings found with GCC 8.3
* enclave: fix byte order conversion of response sender id
* wiresep-keygen reorder output
* guard most memcpy(3) for OOB writes by using MIN


2020-03-18, Version 0.9.1

* fix crash on first packet in client-only mode
* fix an out-of-bound write when parsing an ipv4 listen address
* fix logging unknown destination ip6 addresses
* fix a warning in client-only mode about finding a suitable local port


2020-03-15, Version 0.9.0

* support client-only mode, the listen configuration directive is now optional
* greatly reduce IPC message size
* refactor local address selection, fixes race conditions and stability
* do not exit when a write to the peer socket fails
* shrink socket receive buffers from 524216 to 131054 bytes
* queue at most 50 packets instead of 1000


2019-11-20, Version 0.8.4

* default the user to "_wiresep" if omitted
* append to CFLAGS from environment and drop default -O0 -g


2019-11-18, Version 0.8.3

* don't change the process name, keep wiresep
* let resource limits take the configuration into account
* treat OOM errors in the main loop as transient
* don't notify proxy to destroy unsent sessions


2019-11-14, Version 0.8.2

* replace dropuser with setresuid(2)
* ensure that the log* functions don't change errno
* install example config in /usr/local/share/examples
* don't run makewhatis(8) on make install
* assorted improvements to the documentation


2019-11-03, Version 0.8.1

* avoid compiler error about duplicate enumeration
* debug address family mismatches and reason of inactive sessions
* don't exit on sendwgdatamsg() failure


2019-10-31, Version 0.8.0

* private and pre-shared keys should now be put in a separate file
* automatically look for a private key in /etc/wiresep/tunX.key
* automatically look for an optional pre-shared key in:
    1. /etc/wiresep/global.psk
    2. /etc/wiresep/tunX.psk
    3. /etc/wiresep/tunX.peerY.psk
    4. /etc/wiresep/peerY.psk
* override default paths with new privkeyfile and pskfile config setting
* loosen up permission requirements on config file now that it no longer
  contains any secrets
* set tight resource limits on data and stack size and others
* updates to documentation and manpages
* add wiresep.conf.example
* move the default config file to /etc/wiresep/wiresep.conf
* fix some inconsistencies in parsing the config file


2019-10-13, Version 0.7.0

* completely reimplement session management to improve reliability and
  conformance to the specification
* only malloc after a new session is authenticated and established
* support queueing of multiple packets
* log interface statistics on receiving a USR1 signal
* be silent on startup
* lot's of refinements
* raise status from alpha to beta


2019-08-08, Version 0.6.0

* lot's of small refinements and some code restructuring
* interface public key is no longer needed or allowed, only the private key
* improve DoS resistence in the proxy by looking up sessions in logarithmic time
* use gprof plus a new testing harness to measure the performance of the proxy
* increase socket receive buffers to eight maximum sized UDP packets
* proxy: collect stats and log on SIGUSR1
* improvements to wiresep-keygen(1)


2019-05-05, Version 0.5.0

* first public release


base64.c
/*	$OpenBSD: base64.c,v 1.8 2015/01/16 16:48:51 deraadt Exp $	*/

/*
 * Copyright (c) 2019 Tim Kuijsten
 * Copyright (c) 1996 by Internet Software Consortium.
 *
 * Permission to use, copy, modify, and distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND INTERNET SOFTWARE CONSORTIUM DISCLAIMS
 * ALL WARRANTIES WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES
 * OF MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL INTERNET SOFTWARE
 * CONSORTIUM BE LIABLE FOR ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL
 * DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR
 * PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS
 * ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS
 * SOFTWARE.
 */

/*
 * Portions Copyright (c) 1995 by International Business Machines, Inc.
 *
 * International Business Machines, Inc. (hereinafter called IBM) grants
 * permission under its copyrights to use, copy, modify, and distribute this
 * Software with or without fee, provided that the above copyright notice and
 * all paragraphs of this notice appear in all copies, and that the name of IBM
 * not be used in connection with the marketing of any product incorporating
 * the Software or modifications thereof, without specific, written prior
 * permission.
 *
 * To the extent it has a right to do so, IBM grants an immunity from suit
 * under its patents, if any, for the use, sale or manufacture of products to
 * the extent that such products are used for performing Domain Name System
 * dynamic updates in TCP/IP networks by means of the Software.  No immunity is
 * granted for any product per se or for any other function of any product.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", AND IBM DISCLAIMS ALL WARRANTIES,
 * INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
 * PARTICULAR PURPOSE.  IN NO EVENT SHALL IBM BE LIABLE FOR ANY SPECIAL,
 * DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER ARISING
 * OUT OF OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE, EVEN
 * IF IBM IS APPRISED OF THE POSSIBILITY OF SUCH DAMAGES.
 */

#include <ctype.h>
#include <string.h>

#include "base64.h"

static const char Base64[] =
	"ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";
static const char Pad64 = '=';

/* (From RFC1521 and draft-ietf-dnssec-secext-03.txt)
   The following encoding technique is taken from RFC 1521 by Borenstein
   and Freed.  It is reproduced here in a slightly edited form for
   convenience.

   A 65-character subset of US-ASCII is used, enabling 6 bits to be
   represented per printable character. (The extra 65th character, "=",
   is used to signify a special processing function.)

   The encoding process represents 24-bit groups of input bits as output
   strings of 4 encoded characters. Proceeding from left to right, a
   24-bit input group is formed by concatenating 3 8-bit input groups.
   These 24 bits are then treated as 4 concatenated 6-bit groups, each
   of which is translated into a single digit in the base64 alphabet.

   Each 6-bit group is used as an index into an array of 64 printable
   characters. The character referenced by the index is placed in the
   output string.

                         Table 1: The Base64 Alphabet

      Value Encoding  Value Encoding  Value Encoding  Value Encoding
          0 A            17 R            34 i            51 z
          1 B            18 S            35 j            52 0
          2 C            19 T            36 k            53 1
          3 D            20 U            37 l            54 2
          4 E            21 V            38 m            55 3
          5 F            22 W            39 n            56 4
          6 G            23 X            40 o            57 5
          7 H            24 Y            41 p            58 6
          8 I            25 Z            42 q            59 7
          9 J            26 a            43 r            60 8
         10 K            27 b            44 s            61 9
         11 L            28 c            45 t            62 +
         12 M            29 d            46 u            63 /
         13 N            30 e            47 v
         14 O            31 f            48 w         (pad) =
         15 P            32 g            49 x
         16 Q            33 h            50 y

   Special processing is performed if fewer than 24 bits are available
   at the end of the data being encoded.  A full encoding quantum is
   always completed at the end of a quantity.  When fewer than 24 input
   bits are available in an input group, zero bits are added (on the
   right) to form an integral number of 6-bit groups.  Padding at the
   end of the data is performed using the '=' character.

   Since all base64 input is an integral number of octets, only the
         -------------------------------------------------                       
   following cases can arise:
   
       (1) the final quantum of encoding input is an integral
           multiple of 24 bits; here, the final unit of encoded
	   output will be an integral multiple of 4 characters
	   with no "=" padding,
       (2) the final quantum of encoding input is exactly 8 bits;
           here, the final unit of encoded output will be two
	   characters followed by two "=" padding characters, or
       (3) the final quantum of encoding input is exactly 16 bits;
           here, the final unit of encoded output will be three
	   characters followed by one "=" padding character.
   */

ssize_t
base64_ntop(u_char const *src, size_t srclength, char *target, size_t targsize)
{
	size_t datalength = 0;
	u_char input[3];
	u_char output[4];
	size_t i;

	while (2 < srclength) {
		input[0] = *src++;
		input[1] = *src++;
		input[2] = *src++;
		srclength -= 3;

		output[0] = input[0] >> 2;
		output[1] = ((input[0] & 0x03) << 4) + (input[1] >> 4);
		output[2] = ((input[1] & 0x0f) << 2) + (input[2] >> 6);
		output[3] = input[2] & 0x3f;

		if (datalength + 4 > targsize)
			return (-1);
		target[datalength++] = Base64[output[0]];
		target[datalength++] = Base64[output[1]];
		target[datalength++] = Base64[output[2]];
		target[datalength++] = Base64[output[3]];
	}
    
	/* Now we worry about padding. */
	if (0 != srclength) {
		/* Get what's left. */
		input[0] = input[1] = input[2] = '\0';
		for (i = 0; i < srclength; i++)
			input[i] = *src++;
	
		output[0] = input[0] >> 2;
		output[1] = ((input[0] & 0x03) << 4) + (input[1] >> 4);
		output[2] = ((input[1] & 0x0f) << 2) + (input[2] >> 6);

		if (datalength + 4 > targsize)
			return (-1);
		target[datalength++] = Base64[output[0]];
		target[datalength++] = Base64[output[1]];
		if (srclength == 1)
			target[datalength++] = Pad64;
		else
			target[datalength++] = Base64[output[2]];
		target[datalength++] = Pad64;
	}
	if (datalength >= targsize)
		return (-1);
	target[datalength] = '\0';	/* Returned value doesn't count \0. */
	return (datalength);
}

/* skips all whitespace anywhere.
   converts characters, four at a time, starting at (or after)
   src from base - 64 numbers into three 8 bit bytes in the target area.
   it returns the number of data bytes stored at the target, or -1 on error.
 */

ssize_t
base64_pton(char const *src, u_char *target, size_t targsize)
{
	int state, ch;
	size_t tarindex;
	u_char nextbyte;
	char *pos;

	state = 0;
	tarindex = 0;

	while ((ch = (unsigned char)*src++) != '\0') {
		if (isspace(ch))	/* Skip whitespace anywhere. */
			continue;

		if (ch == Pad64)
			break;

		pos = strchr(Base64, ch);
		if (pos == 0) 		/* A non-base64 character. */
			return (-1);

		switch (state) {
		case 0:
			if (target) {
				if (tarindex >= targsize)
					return (-1);
				target[tarindex] = (pos - Base64) << 2;
			}
			state = 1;
			break;
		case 1:
			if (target) {
				if (tarindex >= targsize)
					return (-1);
				target[tarindex]   |=  (pos - Base64) >> 4;
				nextbyte = ((pos - Base64) & 0x0f) << 4;
				if (tarindex + 1 < targsize)
					target[tarindex+1] = nextbyte;
				else if (nextbyte)
					return (-1);
			}
			tarindex++;
			state = 2;
			break;
		case 2:
			if (target) {
				if (tarindex >= targsize)
					return (-1);
				target[tarindex]   |=  (pos - Base64) >> 2;
				nextbyte = ((pos - Base64) & 0x03) << 6;
				if (tarindex + 1 < targsize)
					target[tarindex+1] = nextbyte;
				else if (nextbyte)
					return (-1);
			}
			tarindex++;
			state = 3;
			break;
		case 3:
			if (target) {
				if (tarindex >= targsize)
					return (-1);
				target[tarindex] |= (pos - Base64);
			}
			tarindex++;
			state = 0;
			break;
		}
	}

	/*
	 * We are done decoding Base-64 chars.  Let's see if we ended
	 * on a byte boundary, and/or with erroneous trailing characters.
	 */

	if (ch == Pad64) {			/* We got a pad char. */
		ch = (unsigned char)*src++;	/* Skip it, get next. */
		switch (state) {
		case 0:		/* Invalid = in first position */
		case 1:		/* Invalid = in second position */
			return (-1);

		case 2:		/* Valid, means one byte of info */
			/* Skip any number of spaces. */
			for (; ch != '\0'; ch = (unsigned char)*src++)
				if (!isspace(ch))
					break;
			/* Make sure there is another trailing = sign. */
			if (ch != Pad64)
				return (-1);
			ch = (unsigned char)*src++;		/* Skip the = */
			/* Fall through to "single trailing =" case. */
			/* FALLTHROUGH */

		case 3:		/* Valid, means two bytes of info */
			/*
			 * We know this char is an =.  Is there anything but
			 * whitespace after it?
			 */
			for (; ch != '\0'; ch = (unsigned char)*src++)
				if (!isspace(ch))
					return (-1);

			/*
			 * Now make sure for cases 2 and 3 that the "extra"
			 * bits that slopped past the last full byte were
			 * zeros.  If we don't check them, they become a
			 * subliminal channel.
			 */
			if (target && tarindex < targsize &&
			    target[tarindex] != 0)
				return (-1);
		}
	} else {
		/*
		 * We ended by seeing the end of the string.  Make sure we
		 * have no partial bytes lying around.
		 */
		if (state != 0)
			return (-1);
	}

	return (tarindex);
}
